state/migrations/20230802121023_device_data_jsonb_test.go

```diff
@@ -0,0 +1,86 @@
+package migrations
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"testing"
+
+	"github.com/jmoiron/sqlx"
+	_ "github.com/lib/pq"
+	"github.com/matrix-org/sliding-sync/internal"
+	"github.com/matrix-org/sliding-sync/testutils"
+)
+
+var postgresConnectionString = "user=xxxxx dbname=syncv3_test sslmode=disable"
+
+func TestMain(m *testing.M) {
+	postgresConnectionString = testutils.PrepareDBConnectionString()
+	exitCode := m.Run()
+	os.Exit(exitCode)
+}
+
+func connectToDB(t *testing.T) (*sqlx.DB, func()) {
+	db, err := sqlx.Open("postgres", postgresConnectionString)
+	if err != nil {
+		t.Fatalf("failed to open SQL db: %s", err)
+	}
+	return db, func() {
+		db.Close()
+	}
+}
+
+func TestJSONBMigration(t *testing.T) {
+	ctx := context.Background()
+	db, close := connectToDB(t)
+	defer close()
+
+	// Create the table in the old format (data = BYTEA instead of JSONB)
+	_, err := db.Exec(`CREATE TABLE syncv3_device_data (
+		user_id TEXT NOT NULL,
+		device_id TEXT NOT NULL,
+		data BYTEA NOT NULL,
+		UNIQUE(user_id, device_id)
+	);`)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tx.Commit()
+
+	// insert some "invalid" data
+	dd := internal.DeviceData{
+		DeviceLists: internal.DeviceLists{
+			New:  map[string]int{"@💣:localhost": 1},
```

Comment:
It's not the unicode that's the problem on its own I think. It's the escaping part in postgres that it fails on. Can you check that this string does fail when escaped?

Reply:
We're not using any escaping now, it is just an issue with the migration itself